### PR TITLE
[Model] Fish Audio S2 Integration

### PR DESCRIPTION
**Pull request by zhaochenyang20**
This PR is adapted from PR #133 . We first merge the branch shuai-talker into the S2 integration PR. Since we found the performance regression of the branch feat/fishaudio_s1_s2pro, we opened a new branch. And, for the features like CUDA Graph and batched codebook decoding, let's make the following PR after this is merged. Make PRs small. Though this PR is quite huge.

For benchmarking:

### Docker

```bash
docker pull frankleeeee/sglang-omni:dev

docker run -it --shm-size 32g --gpus all -v /data/cache/huggingface:/root/.cache/huggingface -v /data/<your-name>:/data --ipc=host --privileged --name sglang-omni-<your-name> frankleeeee/sglang-omni:dev /bin/zsh
```

### Install sglang-omni (inside Docker)

before clone, validate your github.

```bash
git clone git@github.com:sgl-project-dev/sglang-omni.git
cd sglang-omni
git checkout s2-integration-Jing

deactivate
rm -rf .venv
uv venv .venv -p 3.12
source .venv/bin/activate
uv pip install msgpack
uv pip install -e ".[s2pro]"

uv pip install https://github.com/Dao-AILab/flash-attention/releases/download/v2.8.3/flash_attn-2.8.3+cu12torch2.9cxx11abiTRUE-cp312-cp312-linux_x86_64.whl


# Checkpoint loading to be done after fish audio team's hf release
export S2PRO_CKPT=/root/.cache/huggingface/s2-pro/s2-pro
```
### Web-Demo

```bash
uv pip install gradio
./playground/tts/start.sh
```

### Benchmark 

1. seed-tts-eval testset (pre-cached in radixark03)

```bash
export S2PRO_CKPT=/root/.cache/huggingface/s2-pro/s2-pro
tar xzf /root/.cache/huggingface/seed_tts_eval_testset.tar.gz -C /tmp
export SEED_TTS=/tmp
source .venv/bin/activate
uv pip install torchaudio jiwer scipy tqdm
```

2. Performance

These benchmarks launch the server separately. No need to keep the previous server.

```bash
python benchmarks/profile_s2pro_sglang.py \
    --checkpoint $S2PRO_CKPT \
    --testset $SEED_TTS/seedtts_testset/en/meta.lst \
    --output-dir results/s2pro_sglang \
    --max-samples 50 --batch-sizes 1 --save-audio
```

Example results:

```bash
================================================================
  Single Request
================================================================
  Requests:        50
  TTFT mean:       24.7ms (median 24.0ms)
  TTFB mean:       186.5ms (median 183.6ms)
  TTFB p95:        200.4ms
  Tok/s (per-req): 60.4 mean, 60.6 median
  Tok/s (agg):     60.4
  Gen tokens:      84 mean, 4194 total
  Latency mean:    1.388s
  RTF mean:        0.3567
================================================================
```

3. Batched performance

```bash
python benchmarks/profile_s2pro_sglang.py \
    --checkpoint $S2PRO_CKPT \
    --testset $SEED_TTS/seedtts_testset/en/meta.lst \
    --output-dir results/s2pro_sglang_batched \
    --max-samples 50 --batch-sizes 1,2,4,8 --save-audio
```

Example results:

```bash
================================================================
  Single Request
================================================================
  Requests:        50
  TTFT mean:       23.4ms (median 23.3ms)
  TTFB mean:       179.8ms (median 178.1ms)
  TTFB p95:        188.0ms
  Tok/s (per-req): 61.8 mean, 61.9 median
  Tok/s (agg):     61.9
  Gen tokens:      84 mean, 4195 total
  Latency mean:    1.355s
  RTF mean:        0.3483
================================================================

================================================================
  Batch Size 2
================================================================
  Requests:        50
  TTFT mean:       22.0ms (median 22.0ms)
  TTFB mean:       242.6ms (median 225.2ms)
  TTFB p95:        226.9ms
  Tok/s (per-req): 45.7 mean, 46.2 median
  Tok/s (agg):     45.7
  Gen tokens:      81 mean, 4060 total
  Latency mean:    1.776s
  RTF mean:        0.4724
================================================================

================================================================
  Batch Size 4
================================================================
  Requests:        48
  TTFT mean:       31.6ms (median 31.4ms)
  TTFB mean:       319.7ms (median 319.7ms)
  TTFB p95:        321.5ms
  Tok/s (per-req): 31.9 mean, 31.9 median
  Tok/s (agg):     31.9
  Gen tokens:      86 mean, 4108 total
  Latency mean:    2.683s
  RTF mean:        0.6758
================================================================

================================================================
  Batch Size 8
================================================================
  Requests:        48
  TTFT mean:       53.0ms (median 50.2ms)
  TTFB mean:       518.5ms (median 507.7ms)
  TTFB p95:        573.6ms
  Tok/s (per-req): 19.7 mean, 19.7 median
  Tok/s (agg):     19.7
  Gen tokens:      85 mean, 4096 total
  Latency mean:    4.338s
  RTF mean:        1.0956
================================================================
```

4. Results (seed-tts-eval EN, 50 samples, per-request averages)

| Metric | BS=1 | BS=2 | BS=4 | BS=8 |
|---|---|---|---|---|
| tok/s (mean) | 63.3 | 45.8 | 31.9 | 19.6 |
| RTF (mean) | 0.340 | 0.473 | 0.676 | 1.097 |
| Latency (mean) | 1.33s | 1.80s | 2.69s | 4.36s |
| TTFT (mean) | 19.6 ms | 22.0 ms | 31.6 ms | 50.7 ms |
| TTFB (mean) | 172.8 ms | 249.9 ms | 319.1 ms | 509.6 ms |
| Success rate | 50/50 | 50/50 | 48/48 | 48/48 |

5.  Quality / WER

```bash
# Generate audio first (see Performance above), then evaluate WER:
python benchmarks/eval_wer.py \
    --meta $SEED_TTS/seedtts_testset/en/meta.lst \
    --audio-dir results/s2pro_sglang/audio --lang en
```

Example results:

```bash
====================================================
  WER Evaluation Results (EN)
====================================================
  Evaluated:           50/1088
  WER mean:            0.0122 (1.22%)
  WER median:          0.0000
  WER (excl >50%):     0.0122 (1.22%)
  Samples >50% WER:    0 (0.0%)

  Output: results/s2pro_sglang/wer.json
====================================================
```

6. Results

| Metric | Value |
|---|---|
| Samples evaluated | 50 |
| WER (mean) | 0.99% |
| WER (median) | 0.0% |
| WER (p95) | 4.2% |

### CODE REVIEW

File: sglang_omni/models/fishaudio_s2_pro/fish_speech/models/text2semantic/modeling.py [hunk 1]
```diff
@@ -0,0 +1,2466 @@
...
+
+    def forward(self, x: Tensor) -> Tensor:
+        x1, x3 = self.w1(x), self.w3(x)
+        if FISH_BATCH_INVARIANT is False:
+            return self.w2(F.silu(x1) * x3)
+
+        return self.w2(F.silu(x1) * x3)
+
+
+class MoeFusedLinear(nn.Module):
```
Comment (leng-yue) [inline sglang_omni/models/fishaudio_s2_pro/fish_speech/models/text2semantic/modeling.py]: Can also remove this
